### PR TITLE
refactor: Only set `ARGO_TEMPLATE` env for init container.

### LINKED ISSUE #8790: duplicative env var `ARGO_TEMPLATE` result in etcd size 1MB limit
## Summary
I am running a containerSet with lots of steps. And I have large parameter input for the step.
Because there is an env variable called ARGO_TEMPLATE each container, the overall pod description is too large to fit in etcd, exceeding the 1MB limit.
(10 + init + wait) = 12 containers
12 * 100k = 1200k > 1MB
What is the usage of ARGO_TEMPLATE? Can we delete it?

I am using the v3.1.10

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

### PR DESCRIPTION
**Pull request by jswxstw**
Fixes #8790

### Motivation

duplicative env var ARGO_TEMPLATE result in etcd size 1MB limit

### Modifications

Only set `ARGO_TEMPLATE` env for init container, wait container read it from the template file written by init container.

### Verification

existing UT or E2E tests have covered it.

### CODE REVIEW

**Comment by tooptoop4** [comment]
@jswxstw this is great but can u make this change enabled via an environment variable (ARGO_TEMPLATE_ONLY_ON_INIT)? @juliev0 mentioned one of their workflows in main container uses the variable: https://github.com/argoproj/argo-workflows/issues/8790#issuecomment-1743614552 whereas 99% of users (including me) won't and would benefit from your PR

**Comment by juliev0** [comment]
> @jswxstw this is great but can u make this change enabled via an environment variable (ARGO_TEMPLATE_ONLY_ON_INIT)? @juliev0 mentioned one of their workflows in main container uses the variable: [#8790 (comment)](https://github.com/argoproj/argo-workflows/issues/8790#issuecomment-1743614552) whereas 99% of users (including me) won't and would benefit from your PR

Thanks for noticing that @tooptoop4 ! I believe this is one motivation for your making [this](https://github.com/argoproj/argo-workflows/pull/13742/files) configurable, right?

@jswxstw will it make the code too ugly to do that? I'll admit that my use of that environment variable was a hacky workaround for exceeding the maximum size of an argument, as opposed to a promise that Argo Workflows makes will exist. And admittedly, I was later in a separate situation in which for a different Workflow even this hack didn't work, as the environment variable size made the overall Container size too big as I recall. If needed, I can adjust our Workflow so it doesn't try to do this at all. I guess the only question is whether there's anyone else out there for which this will break.

**Comment by jswxstw** [comment]
> @jswxstw will it make the code too ugly to do that?

@juliev0 Agree, I think we should exercise restraint in adding environment variables, as it could increase complexity and confuse users.
Also, #12325 has supported `ARGO_TEMPLATE` env offload by default, it will also affect users who use this env. @tooptoop4 
I do think `ARGO_TEMPLATE` is an internal environment variable, which is not documented and users should not rely on it.

File: docs/environment-variables.md [hunk 1]
```diff
@@ -23,7 +23,7 @@ This document outlines environment variables that can be used to customize behav
...
 | `ARGO_AGENT_MEMORY_LIMIT`                | `resource.Quantity` | `256m`                                                                                      | Memory resource limit for the agent.                                                                                                                                                                                                                                     |
 | `ARGO_POD_STATUS_CAPTURE_FINALIZER`      | `bool`              | `false`                                                                                     | The finalizer blocks the deletion of pods until the controller captures their status.
-| `ARGO_TEMPLATE_WITH_INPUTS_PARAMETERS`      | `bool`              | `true`                                                                                     | Whether to keep inputs.parameters inside the ARGO_TEMPLATE environment variable of pods.
+| `ARGO_TEMPLATE_WITH_INPUTS_PARAMETERS`   | `bool`              | `true`                                                                                      | Whether to keep `inputs.parameters` inside the `ARGO_TEMPLATE` environment variable of pods.
```
Comment (juliev0) [inline docs/environment-variables.md:26]: if we decide that we don't want to support anyone like me who was using ${ARGO_TEMPLATE} for input parameters anymore (per [this](https://github.com/argoproj/argo-workflows/pull/13761#issuecomment-2415589211) comment), then does it mean we should no longer have this environment variable? (I believe @tooptoop4 was trying to support backward compatibility for me and anyone like me, if there is anyone with this being an environment variable)

Comment (jswxstw) [reply docs/environment-variables.md]: Agree, `ARGO_TEMPLATE_WITH_INPUTS_PARAMETERS` removed.

**Review by tooptoop4 — APPROVED** [review]

**Review by agilgur5 — COMMENTED** [review]
So this is no longer a feature (unlike #13742, which added an env var) as it doesn't add anything; could be a `fix` or `refactor` depending on if we want to auto-backport it

**Comment by juliev0** [comment]
Hey @shuangkun - do you want to review this since you added [this](https://github.com/argoproj/argo-workflows/pull/12325) PR which it's based on?

**Comment by tooptoop4** [comment]
@jswxstw 

when i tried to patch this got below error in the wait container:
```

time="2024-10-17T01:57:33 UTC" level=info msg="Starting Workflow Executor" version=v3.4.11
panic: unexpected end of JSON input
goroutine 1 [running]:
github.com/argoproj/argo-workflows/v3/cmd/argoexec/commands.checkErr({0x3517da0, 0xc000011a40})
        /go/src/github.com/argoproj/argo-workflows/cmd/argoexec/commands/root.go:146 +0x6b
github.com/argoproj/argo-workflows/v3/cmd/argoexec/commands.initExecutor()
        /go/src/github.com/argoproj/argo-workflows/cmd/argoexec/commands/root.go:101 +0x3b5
github.com/argoproj/argo-workflows/v3/cmd/argoexec/commands.waitContainer({0x3547bb8, 0xc0009a3e40})
        /go/src/github.com/argoproj/argo-workflows/cmd/argoexec/commands/wait.go:28 +0x4f
github.com/argoproj/argo-workflows/v3/cmd/argoexec/commands.NewWaitCommand.func1(0xc0008dbb80?, {0x2d958d7?, 0x0?, 0x0?})
        /go/src/github.com/argoproj/argo-workflows/cmd/argoexec/commands/wait.go:18 +0x2a
github.com/spf13/cobra.(*Command).execute(0xc0008dbb80, {0x4cda578, 0x0, 0x0})
        /go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:876 +0x67b
github.com/spf13/cobra.(*Command).ExecuteC(0xc0008da000)
        /go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:990 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
        /go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:918
github.com/spf13/cobra.(*Command).ExecuteContext(0x3546178?, {0x3547bb8?, 0xc0009a3e40?})
        /go/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:911 +0x4a
main.main()
        /go/src/github.com/argoproj/argo-workflows/cmd/argoexec/main.go:21 +0x9e
```

**Comment by jswxstw** [comment]
> time="2024-10-17T01:57:33 UTC" level=info msg="Starting Workflow Executor" version=v3.4.11

It seems you are using the argoexec image of old version: v3.4.11. Not updated? @tooptoop4

**Comment by tooptoop4** [comment]
@jswxstw what command do u use to build the controller locally?

**Comment by jswxstw** [comment]
@tooptoop4 I used `make start UI=true`, like [developing locally](https://argo-workflows.readthedocs.io/en/stable/running-locally/#developing-locally).
Please note this, `make argoexec-image`  will build latest image and import it into the k3d cluster:`k3s-default` by default.
QQ_1729146706308

**Comment by jswxstw** [comment]
> So this is no longer a feature (unlike #13742, which added an env var) as it doesn't add anything; could be a `fix` or `refactor` depending on if we want to auto-backport it

@agilgur5 Agree, this PR should be a `refactor`, based on the feature [configmap offloading](https://github.com/argoproj/argo-workflows/pull/12325).

**Review by terrytangyuan — COMMENTED** [review]
This seems like a breaking change. I remember Intuit or other vendors depend on this. @juliev0 Do you remember?

**Comment by tooptoop4** [comment]
> This seems like a breaking change. I remember Intuit or other vendors depend on this. @juliev0 Do you remember?

accepted in https://github.com/argoproj/argo-workflows/pull/13761#discussion_r1802282712

**Comment by agilgur5** [comment]
> This seems like a breaking change.

It's also an undocumented, internal environment variable, so shouldn't be relied on as mentioned [by](https://github.com/argoproj/argo-workflows/pull/13761#issuecomment-2415589211) @jswxstw and I [above](https://github.com/argoproj/argo-workflows/pull/13761#pullrequestreview-2373075241)

**Comment by juliev0** [comment]
Thanks for keeping that in mind, though, @terrytangyuan!

**Review by terrytangyuan — APPROVED** [review]
Thanks!

**Comment by shuangkun** [comment]
> Hey @shuangkun - do you want to review this since you added [the configmap offloading PR](https://github.com/argoproj/argo-workflows/pull/12325) which it's based on?

Sorry, I just found time now. This PR has been merged and it seems there is no problem.

**Pull request merged**